lib/SILOptimizer/Mandatory/TFConstExpr.cpp

```diff
@@ -233,14 +238,14 @@ namespace {
       calculatedValues.insert({ value, symVal });
     }
 
-    void setAddress(SILValue addr, AddressValue addrVal) {
-      assert(addr->getType().isAddress() && "values are tracked separately");
-      addressValues.insert({ addr, addrVal });
-    }
-
+    /// Invariant: Before the call, `addressValues` must not contain `addr` as a
+    /// key, unless `initialValue` is an unknown value, representing a failure in
+    /// compiler-time evaluation.
     AddressValue createMemoryObject(SILValue addr, SymbolicValue initialValue) {
+      if (!initialValue.isUnknown())
+        assert(!addressValues.count(addr));
```

Comment:
Please use:
assert(initialValue.isUnknown() || !addressValues.count(addr));

That said why is this getting unknown initial values?

Reply:
A test case (simplified from `crashers.swift`) is:
```swift
public func testStraightLineXORTraining() {
  // Training data
  let inputBatch = Tensor<Float>([[0.0]])
}
```

The SIL snippet being processed in const eval as part of deabstraction pass is:
```
  %91 = alloc_stack $Float                        // users: %94, %95, %92
  copy_addr %86 to [initialization] %91 : $*Float // id: %92
```

There, `createMemoryObject()` is first called on the `alloc_stack` inst above, and then called again on the same addr value `%91` when procesing the `copy_addr`. It seems we don't yet have code to deal with the latter, so it called into `getUnknownAddress()` at the end of `computeSingleStoreAddressValue()`.

This can be reproduced if we replace this assert above with:
```
assert(!addressValues.count(addr));
```

The full call stack when crashing is:
```
swiftc: /usr/local/google/home/hongm/ssd_part/git/swift-base/swift/lib/SILOptimizer/Mandatory/TFConstExpr.cpp:226: void (anonymous namespace)::ConstExprFunctionState::setAddress(swift::SILValue, (anonymous namespace)::AddressValue): Assertion `!addressValues.count(addr)' failed.
#0 0x0000000004bc244f llvm::sys::PrintStackTrace(llvm::raw_ostream&) /usr/local/google/home/hongm/ssd_part/git/swift-base/llvm/lib/Support/Unix/Signals.inc:487:13
#1 0x0000000004bc08d2 llvm::sys::RunSignalHandlers() /usr/local/google/home/hongm/ssd_part/git/swift-base/llvm/lib/Support/Signals.cpp:68:18
#2 0x0000000004bc2752 SignalHandler(int) /usr/local/google/home/hongm/ssd_part/git/swift-base/llvm/lib/Support/Unix/Signals.inc:350:1
#3 0x00007f40f7e720c0 __restore_rt (/lib/x86_64-linux-gnu/libpthread.so.0+0x110c0)
#4 0x00007f40e40fefcf gsignal (/lib/x86_64-linux-gnu/libc.so.6+0x32fcf)
#5 0x00007f40e41003fa abort (/lib/x86_64-linux-gnu/libc.so.6+0x343fa)
#6 0x00007f40e40f7e37 (/lib/x86_64-linux-gnu/libc.so.6+0x2be37)
#7 0x00007f40e40f7ee2 (/lib/x86_64-linux-gnu/libc.so.6+0x2bee2)
#8 0x000000000120be01 (anonymous namespace)::ConstExprFunctionState::setAddress(swift::SILValue, (anonymous namespace)::AddressValue) /usr/local/google/home/hongm/ssd_part/git/swift-base/swift/lib/SILOptimizer/Mandatory/TFConstExpr.cpp:0:7
#9 0x000000000120bc3c (anonymous namespace)::ConstExprFunctionState::createMemoryObject(swift::SILValue, swift::SymbolicValue) /usr/local/google/home/hongm/ssd_part/git/swift-base/swift/lib/SILOptimizer/Mandatory/TFConstExpr.cpp:253:7
#10 0x000000000120bb28 (anonymous namespace)::ConstExprFunctionState::getUnknownAddress(swift::SILValue, swift::UnknownReason) /usr/local/google/home/hongm/ssd_part/git/swift-base/swift/lib/SILOptimizer/Mandatory/TFConstExpr.cpp:259:14
#11 0x000000000120ae58 (anonymous namespace)::ConstExprFunctionState::computeSingleStoreAddressValue(swift::SILValue) /usr/local/google/home/hongm/ssd_part/git/swift-base/swift/lib/SILOptimizer/Mandatory/TFConstExpr.cpp:1128:12
#12 0x000000000120a063 (anonymous namespace)::ConstExprFunctionState::getAddressValue(swift::SILValue) /usr/local/google/home/hongm/ssd_part/git/swift-base/swift/lib/SILOptimizer/Mandatory/TFConstExpr.cpp:1264:27
#13 0x000000000120666f (anonymous namespace)::ConstExprFunctionState::computeLoadResult(swift::SILValue) /usr/local/google/home/hongm/ssd_part/git/swift-base/swift/lib/SILOptimizer/Mandatory/TFConstExpr.cpp:1142:15
#14 0x00000000012096ed (anonymous namespace)::ConstExprFunctionState::computeCallResult(swift::ApplyInst*) /usr/local/google/home/hongm/ssd_part/git/swift-base/swift/lib/SILOptimizer/Mandatory/TFConstExpr.cpp:866:16
#15 0x000000000120ab9d (anonymous namespace)::ConstExprFunctionState::computeSingleStoreAddressValue(swift::SILValue) /usr/local/google/home/hongm/ssd_part/git/swift-base/swift/lib/SILOptimizer/Mandatory/TFConstExpr.cpp:0:25
#16 0x000000000120a063 (anonymous namespace)::ConstExprFunctionState::getAddressValue(swift::SILValue) /usr/local/google/home/hongm/ssd_part/git/swift-base/swift/lib/SILOptimizer/Mandatory/TFConstExpr.cpp:1264:27
#17 0x000000000120666f (anonymous namespace)::ConstExprFunctionState::computeLoadResult(swift::SILValue) /usr/local/google/home/hongm/ssd_part/git/swift-base/swift/lib/SILOptimizer/Mandatory/TFConstExpr.cpp:1142:15
#18 0x0000000001205e97 (anonymous namespace)::ConstExprFunctionState::computeConstantValue(swift::SILValue) /usr/local/google/home/hongm/ssd_part/git/swift-base/swift/lib/SILOptimizer/Mandatory/TFConstExpr.cpp:383:5
#19 0x0000000001205466 (anonymous namespace)::ConstExprFunctionState::getConstantValue(swift::SILValue) /usr/local/google/home/hongm/ssd_part/git/swift-base/swift/lib/SILOptimizer/Mandatory/TFConstExpr.cpp:962:8
#20 0x0000000001205c5d (anonymous namespace)::ConstExprFunctionState::computeConstantValue(swift::SILValue) /usr/local/google/home/hongm/ssd_part/git/swift-base/swift/lib/SILOptimizer/Mandatory/TFConstExpr.cpp:356:16
#21 0x0000000001205466 (anonymous namespace)::ConstExprFunctionState::getConstantValue(swift::SILValue) /usr/local/google/home/hongm/ssd_part/git/swift-base/swift/lib/SILOptimizer/Mandatory/TFConstExpr.cpp:962:8
#22 0x000000000120525e swift::tf::ConstExprEvaluator::computeConstantValues(llvm::ArrayRef<swift::SILValue>, llvm::SmallVectorImpl<swift::SymbolicValue>&) /usr/local/google/home/hongm/ssd_part/git/swift-base/swift/lib/SILOptimizer/Mandatory/TFConstExpr.cpp:0:25
#23 0x000000000121dcad (anonymous namespace)::TFDeabstraction::checkAttributesAndFormGraphOps() /usr/local/google/home/hongm/ssd_part/git/swift-base/swift/lib/SILOptimizer/Mandatory/TFDeabstraction.cpp:1811:21
#24 0x000000000121c2cb (anonymous namespace)::TFDeabstraction::doIt() /usr/local/google/home/hongm/ssd_part/git/swift-base/swift/lib/SILOptimizer/Mandatory/TFDeabstraction.cpp:2224:3
#25 0x000000000121c096 (anonymous namespace)::TFDeabstractionPass::run() /usr/local/google/home/hongm/ssd_part/git/swift-base/swift/lib/SILOptimizer/Mandatory/TFDeabstraction.cpp:2286:53
#26 0x00000000010a1209 swift::SILPassManager::runModulePass(unsigned int) /usr/local/google/home/hongm/ssd_part/git/swift-base/swift/lib/SILOptimizer/PassManager/PassManager.cpp:474:3
#27 0x00000000010a1bbd swift::SILPassManager::execute() /usr/local/google/home/hongm/ssd_part/git/swift-base/swift/lib/SILOptimizer/PassManager/PassManager.cpp:530:7
#28 0x0000000000711c3b swift::SILPassManager::executePassPipelinePlan(swift::SILPassPipelinePlan const&) /usr/local/google/home/hongm/ssd_part/git/swift-base/swift/include/swift/SILOptimizer/PassManager/PassManager.h:0:7
#29 0x00000000010c3bbd swift::runSILDiagnosticPasses(swift::SILModule&) /usr/local/google/home/hongm/ssd_part/git/swift-base/swift/lib/SILOptimizer/PassManager/Passes.cpp:55:6
#30 0x000000000054e17b performMandatorySILPasses(swift::CompilerInvocation&, swift::SILModule*, swift::FrontendObserver*) /usr/local/google/home/hongm/ssd_part/git/swift-base/swift/lib/FrontendTool/FrontendTool.cpp:1000:9
#31 0x000000000054a499 performCompileStepsPostSILGen(swift::CompilerInstance&, swift::CompilerInvocation&, std::unique_ptr<swift::SILModule, std::default_delete<swift::SILModule> >, bool, llvm::PointerUnion<swift::ModuleDecl*, swift::SourceFile*>, swift::PrimarySpecificPaths const&, bool, int&, swift::FrontendObserver*, swift::UnifiedStatsReporter*) /usr/local/google/home/hongm/ssd_part/git/swift-base/swift/lib/FrontendTool/FrontendTool.cpp:1235:7
#32 0x0000000000546124 performCompile(swift::CompilerInstance&, swift::CompilerInvocation&, llvm::ArrayRef<char const*>, int&, swift::FrontendObserver*, swift::UnifiedStatsReporter*) /usr/local/google/home/hongm/ssd_part/git/swift-base/swift/lib/FrontendTool/FrontendTool.cpp:980:9
#33 0x000000000054502d swift::performFrontend(llvm::ArrayRef<char const*>, char const*, void*, swift::FrontendObserver*) /usr/local/google/home/hongm/ssd_part/git/swift-base/swift/lib/FrontendTool/FrontendTool.cpp:1803:8
#34 0x000000000049b6a4 main /usr/local/google/home/hongm/ssd_part/git/swift-base/swift/tools/driver/driver.cpp:175:7
#35 0x00007f40e40ec2b1 __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x202b1)
#36 0x0000000000493bca _start (/usr/local/google/home/hongm/ssd_part/git/swift-base/build/Ninja-RelWithDebInfoAssert+swift-DebugAssert/swift-linux-x86_64/bin/swiftc+0x493bca)
Stack dump:
0.	Program arguments: /usr/local/google/home/hongm/ssd_part/git/swift-base/build/Ninja-RelWithDebInfoAssert+swift-DebugAssert/swift-linux-x86_64/bin/swiftc -frontend -emit-sil -O test/TensorFlow/sends_recvs_tmp.swift 
1.	While running pass #50 SILModuleTransform "TFDeabstraction".
2.	TFDeabstraction on function $S15sends_recvs_tmp27testStraightLineXORTrainingyyF\0
3.	TFDeabstraction::checkAttributesAndFormGraphOps\0
Aborted
```

Reply:
I see, ok.  It looks like getUnknownAddress() is being abused in computeSingleStoreAddressValue in the failure case.  I agree that making the assertion accept this for now is the right way to go.  Thanks for investigating!